### ISSUE TITLE
Dns66 Is not working in some apps

### ISSUE THREAD

**Issue by Binlabel** [julian-klode/dns66#125]
*UPDATE*

Ads still persist in all of my apps As of 5/26/17.

I uninstalled it weeks ago and decided to try it again recently. No dice still. 

Android version 7.0

Samsung Galaxy s8

Ive tried other methods of blocking ads like adblockplus for android but nothing seems to work.

I think its just my phone/firmware at this point.

**Comment by aggsol** [comment]
Did you check if the uses hostname is blacklisted at all?

**Comment by julian-klode** [comment]
that might just not be an ad served by an ad serving domain.

**Comment by julian-klode** [comment]
1. We can't block all ads
2. To test if DNS66 works, with the default lists, visit eviltracker.net - that should be blocked.

**Comment by Kronzky** [comment]
I also don't see much of an effect in apps. (The eviltracker.net site is blocked ok, but I'm more concerned about ads in apps.)
Here's some I tested, and that seem to be serving Google ads (I've activated Adaway's and Peter Lowe's hosts files):

- [Solid Explorer Classic](https://play.google.com/store/apps/details?id=pl.solidexplorer)
![solidexplorerclassic](https://cloud.githubusercontent.com/assets/11358997/25642939/e7fe1aea-2f6a-11e7-9316-a57b09e69105.jpg)
- [LED Flash Alert](https://play.google.com/store/apps/details?id=com.iterbit.ledactor)
![led_actor](https://cloud.githubusercontent.com/assets/11358997/25642961/0ec54996-2f6b-11e7-880a-93c2929389ce.jpg)
- [Stich & Share](https://play.google.com/store/apps/details?id=com.glitch.stitchandshare)
![stichnshare](https://cloud.githubusercontent.com/assets/11358997/25642987/200a101a-2f6b-11e7-8091-c7cf99ecfb02.jpg)

**Comment by SMed79** [comment]
For "Solid Explorer Classic" try to blacklist that host:  `ads.heyzap.com`

**Comment by julian-klode** [comment]
It should also be noticed that you need to close and open all apps again after you start blocking stuff, otherwise it just keeps the cached ad around.

And @Kronzky having just Peter Lowe's list and the tiny AdAway enabled might not give best results, anyway.

**Comment by Kronzky** [comment]
Yeah, I restarted the apps several times, but it didn't make a difference (even rebooted the phone, but to no avail).
I now activated all the hosts (except for the malware ones), but that only caught **one** of ten apps that I was testing. (When I also tried Adguard yesterday, it killed the ads in _all ten_ of my test apps, so they must be having quite the hosts file...)

**Comment by julian-klode** [comment]
Google ads changed and can't be blocked by just redirecting DNS requests anymore. If you turn AdGuard into the same mode, you see ads as well.

**Comment by riQQ** [comment]
Out of curiosity, why can't Google Ads be blocked by just redirecting DNS requests anymore?

**Comment by julian-klode** [comment]
@riQQ That's a good question, I'd like to know that too. (This only applies to in-app ads, I forgot to mention this). I don't see any DNS requests for the Google ad servers leaving the phone anymore (I forcefully redirect all DNS traffic to my router, and then looked at the log; also redirected all phone traffic to my laptop and looked at it via wireshark), so I can't block them.

**Comment by Kronzky** [comment]
> If you turn AdGuard into the same mode, you see ads as well.

I wasn't quite sure which mode you were referring to, but it seems that, since I don't have a rooted phone, the VPN option is the only one available (with the HTTP proxy one [requiring root](https://adguard.com/en/adguard-android/overview.html)). So they seem to have a better hosts file, if they can still block the app ads, even on an unrooted phone.

**Comment by julian-klode** [comment]
@Kronzky AdGuard has various advanced options that turn on filtering of only DNS requests, compare https://blog.adguard.com/en/great-battery-saving-mode-of-adguard-for-android/ - by default it trades battery life for a better ad blocking experience.

**Comment by Hal17** [comment]
I had some problems a few weeks ago with google ads. 
I read, that the Android VPN function that dns66 is using to register the (manual) DNS servers will switch the whole vpn to use ipv6, if one if the DNS servers has an ipv6 address. 
I use now a few manual DNS servers, only iov4 adresses to avoid that Android will switch the VPN to use ipv6. This works even in WLAN (with dhcp ipv6 DNS server) and in mobile Networks. 
I saw no Google ads since this change of my config. 
I guess that there is "problem" with filtering something in ipv6 world. 
(for couriousity. I had Google ads, in a WLAN working with ipv4, if i used a some ipv4 dns addresses and one (unreachable?)  ipv6 DNS address. The Google ads vanished when I disabled the ipv6 DNS server, and only ipv4 addressed will be allowed) 
So yes. This current config run very well and at the moment i am using adguard no more. Dns66-ipv4 saves my battery. (nice)

**Comment by julian-klode** [comment]
DNS66 0.5.0 (RC1), 0.5.0, and 0.5.1 only enable the networks used by configured DNS servers (and Android blocks other networks). That changed yesterday: IPv6 and IPv4 networking is now always allowed.

Re Your point with "only IPv4" custom servers: That's not true. The service will pick up both your configured servers *and* the ones specified by the connection, so as long as an IPv6 server was present in either set, IPv6 support was enabled for you. Adding more custom servers just makes it less likely that Android falls back to later servers or out of the VPN altogether.

**Comment by nithinkolekar** [comment]
while testing Dns66 with **Merriam Webster dictionary** which is available as offline mode sometimes ads is shown sometime not. Also in some time ad fullscreen popup appears but without any content ie complete blank screen with X button at topleft. If it we can set some log for that particular app we can then identify from which server ad is loading.
Testing is done on non-rooted andoird 5.1.1

**Comment by SMed79** [comment]
merriam site abuse **blob** script, [webrtc](https://en.wikipedia.org/wiki/WebRTC), [websocket](https://en.wikipedia.org/wiki/WebSocket) channels and push their ads from the first party host.

**Comment by alleluia20** [comment]
Today, I met by chance the app "Blokada" at F-Droid, and it seems to block ads in apps nicely (but I have few apps installed, so I could not test it extensively).
I am posting this here because Blokada is open source, so you may want to have a look at the code and get some inspiration...

**Comment by julian-klode** [comment]
Blokada's method there is based on DNS66, and uses the same technology. Strangely enough, not whitelisting system apps by default in the app list (in the top expanding section) makes blocking more accurate in DNS66 as well.

**Comment by julian-klode** [comment]
I believe we should figure out which system apps in particular are responsible for the extra ads and add them to some sort of black list.

**Comment by jeroen7s** [comment]
guessing google routes the google ads dns queries through play services or something like that? not sure

**Comment by mardukeme** [comment]
I figured it out a while ago one late night, but kept forgeting to post here. You need to block two system apps, hopefully these are the two needed if I remembered correctly.  Would be easier if there was a search function to find, and disable apps. 

Ok here are the steps, first need to goto the Apps tab, and than enable Show system apps. Then scroll down till you find these two system apps, and disable them. First is Andriod system webview, and second is html viewer. Last goto the Start tab, and click on the Start button. You should no see the app block ads within apps now.

**Comment by julian-klode** [comment]
OK, I still see some ads in MX Player (the one in the list), but this definitely helped - the full screen ad is no longer shown. Hence closing this bug. If someone finds more system apps we should treat like web browsers (that is, block for), let me know in a new bug.

**Comment by alleluia20** [comment]
Blocking these two system apps does not seem to help much: I get ads in "all" (the two) apps I have with ads: "Map Coordinates" and "Shazam".
Nevertheless, if I remove all the system apps from the whitelist except for Gmail (in order to get the attachments), ads are nicely blocked, and everything looks to work fine. That is, by the way, what Blokada does.
So, why not block all apps by default (except for Gmail, maybe)?

**Comment by julian-klode** [comment]
Broken system apps / OTA updates, essentially.

But yes, ads got worse again since the commit adding webview and friend. So reopening this.

**Comment by aggsol** [comment]
I block all system apps and disable it DNS66 every few weeks to check for updates. A warning message would help the blocking all. F-Droid for example still works perfectly with everything blocked.

**Comment by guest4711** [comment]
Blocking everything (including system apps) is fine for me, too. I block IP traffic of most system apps using AfWall+, too.

**Comment by julian-klode** [comment]
@aggsol Safety first. Some devices completely block UI access if you block system apps. It was a pain to get my Fire TV working again.

**Comment by kevdliu** [comment]
Apologies for replying to such an old thread, but I just want to share that I recently had success blocking more ads in native apps by blocking Google Play Services (I don't want to block all system apps because it might cause a lot of problems as mentioned above). Before blocking play services, I could still see containers for ads with no content, whereas after blocking, all traces of ads are removed. 

Another things I discovered which helped was adding graph.facebook.com to the blacklist as many facebook ads are served through there. The downside to blocking this is that it will break facebook apps since that is also the domain they use for their non-ad APIs. I had to bypass facebook messenger in dns66 for it to function correctly.

**Comment by Zehirmahn** [comment]
Hi, I spoted 3 system apps that need to be toggled down (all 3 or no changes seen in tested apps) :

- Google Backup Transport
- Google Play Services
- Google Services Framework
(In addition to 
  - Android System Webview
  - HTML Viewer
 that are toggled down by default since 0.5.6)

I don't see in app ads anymore but I don't know if it breaks something. Would it be a good idea to have these toggled down by default in a future update?
